kratos/solving_strategies/schemes/residual_based_bdf_custom_scheme.h

```diff
@@ -450,46 +433,56 @@ class ResidualBasedBDFCustomScheme
         const double DeltaTime
         )
     {
         // Values
-        const double dot2un1 = itNode->FastGetSolutionStepValue(Derived2Variable, 1);
-        const double dotun1 = itNode->FastGetSolutionStepValue(DerivedVariable, 1);
-        const double un1 = itNode->FastGetSolutionStepValue(iVar, 1);
-        const double dot2un0 = itNode->FastGetSolutionStepValue(Derived2Variable);
-        double& dotun0 = itNode->FastGetSolutionStepValue(DerivedVariable);
-        double& un0 = itNode->FastGetSolutionStepValue(iVar);
-
-        if (itNode->IsFixed(Derived2Variable)) {
-            dotun0 = (dot2un0 - BDFBaseType::mBDF[1] * dotun1)/BDFBaseType::mBDF[0];
-            un0 = (dotun0 - BDFBaseType::mBDF[1] * un1)/BDFBaseType::mBDF[0];
-        } else if (itNode->IsFixed(DerivedVariable)) {
-            un0 = (dotun1 - BDFBaseType::mBDF[1] * un1)/BDFBaseType::mBDF[0];
-        } else if (itNode->IsFixed(iVar) == false) {
-            un0 = un1 + DeltaTime * dotun1 + 0.5 * std::pow(DeltaTime, 2) * dot2un1;
+        const double dot2un1 = itNode->FastGetSolutionStepValue(rDerived2Variable, 1);
+        const double dotun1 = itNode->FastGetSolutionStepValue(rDerivedVariable, 1);
+        const double un1 = itNode->FastGetSolutionStepValue(rVariable, 1);
+        const double dot2un0 = itNode->FastGetSolutionStepValue(rDerived2Variable);
+        double& dotun0 = itNode->FastGetSolutionStepValue(rDerivedVariable);
+        double& un0 = itNode->FastGetSolutionStepValue(rVariable);
+
+        bool predicted = false;
+
+        if (itNode->HasDofFor(rDerived2Variable)) {
+            if (itNode->IsFixed(rDerived2Variable)) {
+                dotun0 = dot2un0;
+                for (std::size_t i_order = 1; i_order < BDFBaseType::mOrder + 1; ++i_order)
+                    dotun0 -= BDFBaseType::mBDF[i_order] * itNode->FastGetSolutionStepValue(rDerivedVariable, i_order);
+                dotun0 /= BDFBaseType::mBDF[0];
+
+                un0 = dotun0;
+                for (std::size_t i_order = 1; i_order < BDFBaseType::mOrder + 1; ++i_order)
+                    un0 -= BDFBaseType::mBDF[i_order] * itNode->FastGetSolutionStepValue(rVariable, i_order);
+                un0 /= BDFBaseType::mBDF[0];
+                predicted = true;
+            }
         }
-
-        for (std::size_t i_order = 2; i_order < BDFBaseType::mOrder + 1; ++i_order) {
-            const double dotun = itNode->FastGetSolutionStepValue(DerivedVariable, i_order);
-            const double un = itNode->FastGetSolutionStepValue(iVar, i_order);
-
-            if (itNode->IsFixed(Derived2Variable)) {
-                dotun0 -= (BDFBaseType::mBDF[i_order] * dotun)/BDFBaseType::mBDF[0];
-                un0 -= (BDFBaseType::mBDF[i_order] * un)/BDFBaseType::mBDF[0];
-            } else if (itNode->IsFixed(DerivedVariable)) {
-                un0 -= (BDFBaseType::mBDF[i_order] * un)/BDFBaseType::mBDF[0];
+        if (itNode->HasDofFor(rDerivedVariable) && !predicted) {
```

Comment:
minor, but i would put this as

~~~cpp
if (!predicted )
{
     if(itNode->HasDofFor(rDerivedVariable) && itNode->IsFixed(rDerivedVariable))
          ...
     else if (itNode->HasDofFor(rVariable) )
     {
           if ( ! itNode->IsFixed(rVariable))  )
                ...
           else
                ...
~~~

along the same line you could also get rid of the predicted...

i am suggesting this since this way the expensive check (HasDofFor) is done only in the case the predict

Reply:
This is the custom, as I told you this requires extra work to up to date